R/render_release.R

```diff
@@ -35,109 +35,178 @@
-  project_protocols_rel <- path_rel(project_protocols, path(git_root,
-                                                            "src",
-                                                            "project"))
+  git_root <- find_root(is_git_root)
+  output_root <- file.path(git_root, output_root)
 
-  # search for version number
-  all_protocols_index <- c(thematic_protocols_index, project_protocols_index)
-  all_protocols <- c(thematic_protocols, project_protocols)
-  all_protocols_rel <- c(thematic_protocols_rel, project_protocols_rel)
-  version_numbers <- map_chr(all_protocols_index, function(x) {
-    yml <- yaml_front_matter(x)
-    vn <- as.character(yml$params$version_number)
-    vn
+  protocol_index <- dir_ls(
+    file.path(git_root, "src"),
+    type = "file",
+    recurse = 3,
+    regexp = "index\\.Rmd"
+  )
+  yaml <- map(protocol_index, yaml_front_matter)
+  parameters <- map(yaml, "params")
+  version <- map(parameters, "version_number")
+  missing_version <- !map_lgl(version, is.string)
+  if (any(missing_version)) {
+    stop(
+      "version not a string in: ",
+      paste(protocol_index[missing_version], collapse = "; ")
+    )
+  }
+  version <- map_chr(parameters, "version_number")
+  wrong_format <- !grepl("[0-9]{4}\\.[0-9]{2}", version)
+  if (any(wrong_format)) {
+    stop(
+      "version not in YYYY-XX format: ",
+      paste(protocol_index[wrong_format], collapse = "; ")
+    )
+  }
+  protocol_index <- protocol_index[order(version)]
+  yaml <- yaml[order(version)]
+  parameters <- parameters[order(version)]
+  version <- sort(version)
+  for (i in seq_along(protocol_index)) {
+    target_dir <- file.path(output_root, version[i])
+    if (dir_exists(target_dir)) {
+      if (i < length(protocol_index)) {
+        next
+      }
+      dir_delete(target_dir)
     }
+    if (!dir_exists(file.path(dirname(protocol_index[i]), "css"))) {
+      dir_copy(
+        system.file("css", package = "protocolhelper"),
+        file.path(dirname(protocol_index[i]), "css")
+      )
+    }
+    setwd(dirname(protocol_index[i]))
+    render_book(
+      input = ".",
+      output_format = gitbook(
+        split_by = "none",
+        split_bib = FALSE,
+        pandoc_args = c(
+          as.vector(
+            sapply(
+              parameters[[i]][["reviewers"]],
+              pandoc_variable_arg,
+              name = "reviewer"
+            )
+          ),
+          pandoc_variable_arg(
+            "file_manager", parameters[[i]][["file_manager"]]
+          ),
+          pandoc_variable_arg(
+            "protocol_code", parameters[[i]][["protocol_code"]]
+          ),
+          pandoc_variable_arg(
+            "version_number", parameters[[i]][["version_number"]]
+          ),
+          pandoc_variable_arg("thema", parameters[[i]][["theme"]]),
+          pandoc_variable_arg("lang", parameters[[i]][["language"]])
+        ),
+        template = "css/gitbook.html"
+      ),
```

Comment:
Is this necessary? Except for line 118, I think it is already defined in _bookdown.yml and yaml front matter in the index.Rmd files.

Reply:
Yes. This [line](https://github.com/inbo/protocolhelper/blob/882952bf707d3650abcd287719c13f25fe89a03b/R/render_release.R#L58) reads the information the index.Rmd. The `pandoc_variable_arg` stuff provides the information to the [template](https://github.com/inbo/protocolhelper/blob/882952bf707d3650abcd287719c13f25fe89a03b/inst/css/gitbook.html#L5), so it becomes metadata in the HTML. Which is reuse [later](https://github.com/inbo/protocolhelper/blob/882952bf707d3650abcd287719c13f25fe89a03b/R/render_release.R#L166)

Reply:
Nice!